weed/pb/server_discovery.go

```diff
@@ -0,0 +1,59 @@
+package pb
+
+import (
+	"github.com/seaweedfs/seaweedfs/weed/glog"
+)
+
+// ServerDiscovery encodes a way to find at least 1 instance of a service,
+// and provides utility functions to refresh the instance list
+type ServerDiscovery struct {
+	list      []ServerAddress
+	srvRecord *ServerSrvAddress
+}
+
+func NewServiceDiscoveryFromMap(m map[string]ServerAddress) (sd *ServerDiscovery) {
+	sd = &ServerDiscovery{}
+	for _, s := range m {
+		sd.list = append(sd.list, s)
+	}
+	return sd
+}
+
+// RefreshBySrvIfAvailable performs a DNS SRV lookup and updates list with the results
+// of the lookup
+func (sd *ServerDiscovery) RefreshBySrvIfAvailable() {
+	if sd.srvRecord == nil {
+		return
+	}
+	newList, err := sd.srvRecord.LookUp()
+	if err != nil {
+		glog.V(0).Infof("failed to lookup SRV for %s: %v", *sd.srvRecord, err)
+	}
+	if newList == nil || len(newList) == 0 {
+		glog.V(0).Infof("looked up SRV for %s, but found no well-formed names", *sd.srvRecord)
+		return
+	}
+	sd.list = newList
```

Comment:
minor: If the newList is the same as the old list. No need to change the list. 

This may help memory GC a tiny bit, since newList is in stack can be freed faster.

Reply:
👍 didn't know that